src/gt4py/next/_allocators.py

```diff
@@ -323,20 +383,215 @@ def allocate(
             If illegal or inconsistent arguments are specified.
 
     """
-    if device is None and allocator is None:
-        raise ValueError("No 'device' or 'allocator' specified.")
-    actual_allocator = get_allocator(allocator)
-    if actual_allocator is None:
-        assert device is not None  # for mypy
-        actual_allocator = device_allocators[device.device_type]
-    elif device is None:
-        device = core_defs.Device(actual_allocator.__gt_device_type__, 0)
-    elif device.device_type != actual_allocator.__gt_device_type__:
-        raise ValueError(f"Device '{device}' and allocator '{actual_allocator}' are incompatible.")
 
-    return actual_allocator.__gt_allocate__(
-        domain=common.domain(domain),
+    return allocator.__gt_allocate__(
+        domain=domain,
         dtype=dtype,
         device_id=device.device_id,
         aligned_index=aligned_index,
     )
+
+
+def _check_unsupported_device_and_aligned_index(
+    device: Optional[core_defs.Device], aligned_index: Optional[Sequence[common.NamedIndex]]
+) -> None:
+    if aligned_index is not None:
+        raise NotImplementedError("Aligned index is not support for Array API namespaces.")
+    if device is not None:
+        # TODO(havogt): this requires to translate our device object to the concrete Array API implementation's device object
+        raise NotImplementedError("Device specification is not yet supported.")
+
+
+def _get_actual_allocator_and_device(
+    allocator: Optional[FieldBufferAllocationUtil], device: Optional[core_defs.Device]
+) -> tuple[FieldBufferAllocatorProtocol, core_defs.Device]:
+    if allocator is None and device is not None:
+        return device_allocators[device.device_type], device
+
+    actual_allocator = get_allocator(allocator, default=device_allocators[core_defs.DeviceType.CPU])
+    assert actual_allocator is not None
+    if device is None:
+        device = core_defs.Device(actual_allocator.__gt_device_type__, 0)
+    elif device.device_type != actual_allocator.__gt_device_type__:
+        raise ValueError(f"Device '{device}' and allocator '{actual_allocator}' are incompatible.")
+    return actual_allocator, device
+
+
+def get_array_allocation_namespace(
+    allocator: Optional[FieldBufferAllocationUtil | core_defs.ArrayApiNamespace],
+    device: Optional[core_defs.Device] = None,
+) -> GTArrayAllocationNamespace:
```

Comment:
Maybe put aligned_index here and add it as a default to the construction functions.
Same for device.